src/main/java/com/github/dockerjava/core/command/PullImageResultCallback.java

```diff
@@ -19,14 +21,71 @@
 
     private final static Logger LOGGER = LoggerFactory.getLogger(PullImageResultCallback.class);
 
+    private boolean isSwarm = false;
+    private Map<String, PullResponseItem> results = null;
     private PullResponseItem latestItem = null;
 
     @Override
     public void onNext(PullResponseItem item) {
-        this.latestItem = item;
+        // only do it once
+        if (results == null && latestItem == null) {
+            checkForDockerSwarmResponse(item);
+        }
+
+        if (isSwarm) {
+            handleDockerSwarmResponse(item);
+        } else {
+            handleDockerClientResponse(item);
+        }
         LOGGER.debug(item.toString());
     }
 
+    public void checkForDockerSwarmResponse(PullResponseItem item) {
+        if (item.getStatus().matches("Pulling\\s.+\\.{3}$")) {
+            isSwarm = true;
+            LOGGER.debug("Communicating with Docker Swarm.");
+        }
+    }
+
+    public void handleDockerSwarmResponse(PullResponseItem item) {
+        if (results == null) {
+            results = new HashMap<String, PullResponseItem>();
+        }
+        results.put(item.getId(), item);
+    }
+
+    public void checkDockerSwarmPullSuccessful() {
+        if (results.isEmpty()) {
+            throw new DockerClientException("Could not pull image");
+        } else {
+            boolean pullFailed = false;
+            StringBuilder sb = new StringBuilder();
+
+            for (PullResponseItem pullResponseItem : results.values()) {
```

Comment:
How this should work. The first item that doesn't indicate a succesfull pull will lead to pull error.